aiidalab/app.py

```diff
@@ -0,0 +1,647 @@
+        from shutil import rmtree
+        cannot_modify = self.cannot_modify_app()
+
+        # Check if one cannot install the app.
+        if cannot_modify:
+            pass
+        elif self.name == 'home':
+            cannot_modify = "can't remove the home app"
+        elif self.found_local_versions():
+            cannot_modify = "you have local branches"
+        else:
+            # look for the local commited modifications all the available branches
+            initial_value = self.current_version
+            for key, value in self.available_versions.items():
+                self.version.selected.value = value  # switching to the branch in value
+                self._change_version(sleep_time=0)  # actually switching the branch
+                if self.found_local_commits():
+                    cannot_modify = "you have local commits ({} branch)".format(key)
+                    self.version.selected.value = initial_value  # switch back to the initial version
+                    self._change_version(sleep_time=0)  # actually switching the branch
+                    break
+
+        # And finally: uninstall or not?
+        if cannot_modify:
+            self.install_info.value = """<i class="fa fa-times" style="color:red;font-size:4em;" >
+            </i>Can not delete the repository: {}""".format(cannot_modify)
+            sleep(3)
+
+        # Perform uninstall process.
+        else:
+            self.install_info.value = """<i class="fa fa-spinner fa-pulse" style="color:#337ab7;font-size:4em;" ></i>
+            <font size="1"><blink>Unistalling the app...</blink></font>"""
+            sleep(1)
+            rmtree(self._get_appdir())
+            if hasattr(self, '_current_version'):
+                delattr(self, '_current_version')
+            if hasattr(self, '_available_versions'):
+                delattr(self, '_available_versions')
+            self.install_info.value = """<i class="fa fa-check" style="color:#337ab7;font-size:4em;" ></i>
+            <font size="1">Success</font>"""
+            self._refresh_version()
+            self._refresh_install_button()
+            self._refresh_update_button()
+            self._refresh_uninstall_button()
+            sleep(1)
+        self.install_info.value = ''
+
+    def _refresh_uninstall_button(self):
+        """Refresh uninstall button."""
+        if self.is_installed():
+            self.uninstall_button.disabled = False
+            self.uninstall_button.button_style = 'danger'
+        else:
+            self.uninstall_button.disabled = True
+            self.uninstall_button.button_style = ''
+
+    @property
+    def version(self):
+        """App's version."""
+        if not hasattr(self, '_version'):
+            self._version = Version()
+            self._version.change_btn.on_click(self._change_version)
+            self._refresh_version()
+        return self._version
+
+    @property
+    def refs_dict(self):
```

Comment:
This should probably be an internal function.

Reply:
I agree, but let's keep it like this for the moment. If you are going to take git-handling functionality away from the class - I see no point in renaming the function now.